site/content/blog/2020-06-10-opentracing.md

```diff
@@ -0,0 +1,576 @@
+				params = append(params, methodParam{Name: paramName.Name, Type: paramType})
+			}
+		}
+	}
+
+	if e.Results != nil {
+		for _, r := range e.Results.List {
+			typeStr := formatNode(src, r.Type)
+			if len(r.Names) > 0 {
+				for _, k := range r.Names {
+					results = append(results, fmt.Sprintf("%s %s", k.Name, typeStr))
+				}
+			} else {
+				results = append(results, typeStr)
+			}
+		}
+	}
+	return methodData{Params: params, Results: results}
+}
+```
+
+Now we can run the parser on our interface and we'll get something like: `map[Speak:{Params:[{Name:x Type:int}] Results:[]}]`
+As you can see, we collected all the information needed about interface methods and we can now move on to generating decorator with this data.
+
+### Code generation using templates
+
+Let's get to it! We'll start by defining a few helper functions that will be useful during code generation. They will operate on the metadata we've collected before.
+
+```go
+	helperFuncs := template.FuncMap{
+		"joinResults": func(results []string) string {
+			return strings.Join(results, ", ")
+		},
+		"joinResultsForSignature": func(results []string) string {
+			return fmt.Sprintf("(%s)", strings.Join(results, ", "))
+		},
+		"joinParams": func(params []methodParam) string {
+			paramsNames := []string{}
+			for _, param := range params {
+				s := param.Name
+				if strings.HasPrefix(param.Type, "...") {
+					s += "..."
+				}
+				paramsNames = append(paramsNames, s)
+			}
+			return strings.Join(paramsNames, ", ")
+		},
+		"joinParamsWithType": func(params []methodParam) string {
+			paramsWithType := []string{}
+			for _, param := range params {
+				paramsWithType = append(paramsWithType, fmt.Sprintf("%s %s", param.Name, param.Type))
+			}
+			return strings.Join(paramsWithType, ", ")
+		},
+	}
+```
+
+Next we'll create a [Go Template](https://golang.org/pkg/text/template/) for both our decorators:
+
+{{< highlight go "linenos=table" >}}
+	tracerTemplate := `
+	package animals
+
+	type AnimalTracer struct {
+		IAnimal
+	}
+	{{range $index, $element := .}}
+	func (a *AnimalTracer) {{$index}}({{$element.Params | joinParamsWithType}}) {{$element.Results | joinResultsForSignature}} {
+		fmt.Printf("Running {{$index}}({{$element.Params | joinParams}}) with {{range $paramIdx, $param := $element.Params}}'{{$param.Name}}'=%v {{end}}",{{$element.Params | joinParams}})
+		{{- if $element.Results | len | eq 0}}
+			a.IAnimal.{{$index}}({{$element.Params | joinParams}})
+		{{else}}
+			return a.IAnimal.{{$index}}({{$element.Params | joinParams}})
+		{{end}}	
+	}
+	{{end}}
+	`
+{{< / highlight >}}
+
+I know it looks a little scary, but the premise is rather simple. Given the following metadata: `map[Speak:{Params:[{Name:x Type:int}] Results:[]}]` we want to generate a new struct that embeds our `Animal` and wraps it's calls in additional functionality.
+
+I'll go through the template line by line:
+
+**2-6**: Define the new struct
+**7**: Iterate over methods in our metadata
```

Comment:
Separate these by newlines. They actually appear in a single para in the markdown. Or make it a bulleted list, to prevent extra linegaps.